futures-util/src/future/join_all.rs

```diff
@@ -53,30 +61,37 @@ where
 /// # Examples
 ///
 /// ```
-/// use futures_util::future::{FutureExt, join_all, ready};
+/// #![feature(async_await, await_macro, futures_api)]
+/// # futures::executor::block_on(async {
+/// use futures::future::{join_all};
+///
+/// async fn foo(i: u32) -> u32 { i }
+///
+/// let futures = vec![foo(1), foo(2), foo(3)];
 ///
-/// let f = join_all(vec![
-///     ready::<u32>(1),
-///     ready::<u32>(2),
-///     ready::<u32>(3),
-/// ]);
-/// let f = f.map(|x| {
-///     assert_eq!(x, [1, 2, 3]);
-/// });
+/// assert_eq!(await!(join_all(futures)), [1, 2, 3]);
+/// # });
 /// ```
 pub fn join_all<I>(i: I) -> JoinAll<I::Item>
 where
     I: IntoIterator,
     I::Item: Future,
 {
-    let elems = i.into_iter().map(ElemState::Pending).collect();
+    let elems: Vec<_> = i.into_iter().map(ElemState::Pending).collect();
+    let elems = unsafe { Pin::new_unchecked(elems.into_boxed_slice()) };
     JoinAll { elems }
 }
 
+impl<F> JoinAll<F>
+where
+    F: Future,
+{
+    unsafe_pinned!(elems: Pin<Box<[ElemState<F>]>>);
```

Comment:
This isn't necessary-- `JoinAll` and `Pin<Box<...>>` are both `Unpin`, so it's safe to access `elems` directly.

Reply:
uh, yeah, don't know what I was thinking